File: crack-data-structures-and-algorithms/leetcode/reverse_integer_q7.py
# 核心思路
# 对于负数，需要首先转换成正数，否则默认floor取整的除法会导致 -1 / 10 = -1这种情况的出现
# 再注意一下初始val=0的更新条件
class Solution(object):
    def reverse(self, x):
        """
        :type x: int
        :rtype: int
        """
        val = 0

        sign = False
        if x < 0:
            sign = True
            x = -x

        while x != 0:
            r = x % 10
            x /= 10
            if val == 0:
                val = r
            else:
                val = val * 10 + r

        if sign:
            val = -val

        # Handling overflow case.
        INT_MIN = -(1 << 31)
        INT_MAX = (1 << 31) - 1
        if val < INT_MIN:
            val = 0
        elif val > INT_MAX:
            val = 0

        return val

File: crack-data-structures-and-algorithms/leetcode/reverse_integer_q7.cpp
#include <algorithm>
#include <cstdint>

// 核心思路
// 为了处理 int32 的范围问题，转换成 int64_t
// 并且不需要特别处理负数，直接常规/10以及%10分离出每一位即可
class Solution {
public:
    int reverse(int x) {
        int64_t val = static_cast<int64_t>(x);
        int64_t result = 0;
        while (val != 0) {
            auto m = val % 10;
            val /= 10;
            result = result * 10 + m;
        }

        return INT_MIN <= result && result <= INT_MAX ? result : 0;
    }
};
